src/graphics/vertex-format.js

```diff
@@ -49,10 +57,101 @@ Object.assign(pc, function () {
+     *
+     * * {@link pc.SEMANTIC_POSITION}
+     * * {@link pc.SEMANTIC_NORMAL}
+     * * {@link pc.SEMANTIC_TANGENT}
+     * * {@link pc.SEMANTIC_BLENDWEIGHT}
+     * * {@link pc.SEMANTIC_BLENDINDICES}
+     * * {@link pc.SEMANTIC_COLOR}
+     * * {@link pc.SEMANTIC_TEXCOORD0}
+     * * {@link pc.SEMANTIC_TEXCOORD1}
+     * * {@link pc.SEMANTIC_TEXCOORD2}
+     * * {@link pc.SEMANTIC_TEXCOORD3}
+     * * {@link pc.SEMANTIC_TEXCOORD4}
+     * * {@link pc.SEMANTIC_TEXCOORD5}
+     * * {@link pc.SEMANTIC_TEXCOORD6}
+     * * {@link pc.SEMANTIC_TEXCOORD7}
+     *
+     * If vertex data has a meaning other that one of those listed above, use the user-defined
+     * semantics: pc.SEMANTIC_ATTR0 to pc.SEMANTIC_ATTR15.
+     */
+    /**
+     * @name pc.VertexAttributeElement#numComponents
+     * @type {number}
+     * @description The number of components of the vertex attribute.
+     * Can be 1, 2, 3 or 4.
+     */
+    /**
+     * @name pc.VertexAttributeElement#dataType
+     * @type {number}
+     * @description The data type of the attribute. Can be:
      *
-     * @param {boolean} [description[].normalize] - If true, vertex attribute data will be mapped from a
+     * * {@link pc.TYPE_INT8}
+     * * {@link pc.TYPE_UINT8}
+     * * {@link pc.TYPE_INT16}
+     * * {@link pc.TYPE_UINT16}
+     * * {@link pc.TYPE_INT32}
+     * * {@link pc.TYPE_UINT32}
+     * * {@link pc.TYPE_FLOAT32}
+     */
+    /**
+     * @name pc.VertexAttributeElement#[normalize]
+     * @type {boolean}
+     * @description If true, vertex attribute data will be mapped from a
      * 0 to 255 range down to 0 to 1 when fed to a shader. If false, vertex attribute data is left
      * unchanged. If this property is unspecified, false is assumed.
+     */
+    /**
+     * @name pc.VertexAttributeElement#offset
+     * @type {number}
+     * @description The number of initial bytes at the start of a vertex that are not relevant to this attribute.
+     */
+    /**
+     * @name pc.VertexAttributeElement#stride
+     * @type {number}
+     * @description The number of total bytes that are between the start of one vertex, and the start of the next.
+     */
+    /**
+     * @name pc.VertexAttributeElement#scopeId
+     * @type {pc.ScopeId}
+     * @description The shader input variable corresponding to the attribute.
+     */
+    /**
+     * @name pc.VertexAttributeElement#size
+     * @type {number}
+     * @description The size of the attribute in bytes.
+     */
+    var VertexAttributeElement = function () {};
+
+    /**
+     * @class
+     * @name pc.VertexFormat
+     * @classdesc A vertex format is a descriptor that defines the layout of vertex data inside
+     * a {@link pc.VertexBuffer}.
+     * @description Returns a new pc.VertexFormat object.
+     * @param {pc.GraphicsDevice} graphicsDevice - The graphics device used to manage this vertex format.
+     * @param {pc.VertexAttributeDescription[]} description - An array of vertex attribute descriptions.
```

Comment:
I'm a bit confused here. If the `pc.VertexFormat` constructor takes an array of `pc.VertexAttributeDescription` objects, then I would expect a PlayCanvas code example to be something like:

```
var position = new pc.VertexAttributeDescription(pc.SEMANTIC_POSITION, 3, pc.TYPE_FLOAT32);
var uv = new pc.VertexAttributeDescription(pc.SEMANTIC_TEXCOORD0, 2, pc.TYPE_FLOAT32);
var color = new pc.VertexAttributeDescription(pc.SEMANTIC_COLOR, 4, pc.TYPE_UINT8. true);
var vertexFormat = new pc.VertexFormat(graphicsDevice, [ position, uv, color ]);
```

But it's still just using vanilla Object instances, right?

Reply:
Both `pc.VertexAttributeDescription` and `pc.VertexAttributeElement` are not classes but interfaces. Meaning they are primarily used for documenting (the properties of each object) and provide typings for TypeScript. The interfaces don't actually need to be referenced by any user code.

Reply:
But how will the user know the difference when reading the API ref?

Reply:
The description of each begins with saying it's an interface. However, you have a point, the API jsdoc template doesn't clearly differentiate between classes and interfaces. So I've added a proper constructor to each interface, with all relevant parameters. So that if a dev "accidentally" uses one of the interfaces as a class it will still work as expected.

Reply:
@willeastcott Please let me know if you have any remaining concerns.

Reply:
Sorry for the delay in responding. I hope I don't come across as annoying, but I'm still not getting why this PR is useful. I just did a test with `tsc` and it compiles current `pc.VertexFormat` constructor calls just fine. As of today, the second param for the constructor is nice and succinct, using a simple array of JS objects. This PR effectively introduces a second valid way of calling the constructor (which I feel is confusing for users reading the API ref and browsing existing project sourcebases). This new way of calling the constructor is verbose (`new pc.VertexAttributeDescription` repeated for potentially several attributes). Plus, the PR adds 230 lines to the engine codebase - and I'm struggling to see the benefit.

Reply:
@willeastcott Not annoying at all :) I fully understand why it may not be obvious. Let me explain:

If you look closely at the definition of the `pc.VertexFormat` constructor in `playcanvas.d.ts` you will see that the `description` parameter type is incorrectly interpreted as an array of unknown `object`, instead of an array of a well-defined type (i.e. `{ semantic,  components, type, normalize }`):

![image](https://user-images.githubusercontent.com/255073/77848188-66792c00-71c3-11ea-9cc2-66db6d7a3f86.png)


This is because current jsdoc comments use `description[].xxx` to define the properties of the type. It would be fine if `description` was not an array, but a single element. Unfortunately, `tsd-jsdoc` only supports the later case. So either we add support for it in `tsd-jsdoc` which is a lot more complex and uncertain endeavor or we simply move out the jsdoc comments into a separate type, i.e. `pc.VertexAttributeDescription`. Because in the current state, `description` has no proper typings. The later solution has the side benefit of allowing developer code in TypeScript to also reference the element type without having to redefine it (when for example passing on the parameter to PlayCanvas).

Reply:
So basically this is a matter of getting stricter type checking for Typescript developers when specifying a vertex format description at the cost of adding complexity to the codebase and the API reference. It's not a straightforward win IMHO. I'm actually very sensitive to adding new classes to the sidebar of the API reference manual. It's already massively long and if anything, I want to simplify it. And my problem about it confusing developers is still valid. By adding it to the API ref, we're saying "Yeah, you _can_ also do it this way. You don't have to. But this class is actually just here to make type checking more rigorous". It may well be a lot harder to get this addressed in tsd-jsdoc but that _is_ where the problem lies - not the PlayCanvas Engine API.

Reply:
@willeastcott Fair enough. I respect your desire to keep the API classes to a minimum.

What about `VertexAttributeElement`? Unlike parameter `description`, `pc.VertexFormat#elements` is a property and there is [no documented valid way in jsdoc to define array-element-properties using `@property`](https://jsdoc.app/tags-property.html), similar to `description[].xxx`.

Additionally, constructor `VertexIteratorAccessor` takes a `VertexAttributeElement`, which means without this new type we would have to "embed" an anonymous definition (equivalent to `VertexAttributeElement`) twice in two different files and remember to keep them in sync, instead of having everything in one place.